### PR TITLE
Add mockUserToken support for Firestore Emulator.

### PR DESCRIPTION
**Pull request by yuchenshi**
The Firestore version of #4792.

See tests added. Also manually tested end-to-end with modified security rules quickstart.

Fixes #4715.

### CODE REVIEW

File: .changeset/lemon-ligers-protect.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+"@firebase/app-types": minor
+"@firebase/app": minor
+"@firebase/firestore-types": minor
+"@firebase/firestore": minor
```
Comment (Feiyang1) [inline .changeset/lemon-ligers-protect.md:5]:
```suggestion
"@firebase/firestore": minor
"firebase": minor
```

File: packages/app-types/index.d.ts [hunk 2]
```diff
@@ -121,6 +121,84 @@ export interface VersionService {
   version: string;
 }
 
+export type FirebaseSignInProvider =
```
Comment (Feiyang1) [inline packages/app-types/index.d.ts:124]: These types are defined in `@firebase/util` in https://github.com/firebase/firebase-js-sdk/pull/4792. Can we just reuse them?

**Review by Feiyang1 — COMMENTED** [review]
To fix https://github.com/firebase/firebase-js-sdk/discussions/4715, does it mean developers need to rewrite their code to use firestore APIs directly:
```
const firestore = getFirestore();
useFirestoreEmulator(..., {mockUserToken});
```
instead of relying on `@firebase/rules-unit-testing` to create a firestore instance configured with emulator ?

What is the plan to update `@firebase/rules-unit-testing` the library itself and the documentation?

File: packages/app-types/index.d.ts (hunk 2 above)
Comment (yuchenshi) [reply packages/app-types/index.d.ts:124]: Is it okay to introduce a new `@firebase/app-types` -> `@firebase/util` dependency for this?

**Comment by yuchenshi** [comment]
> To fix #4715, does it mean developers need to rewrite their code to use firestore APIs directly:
> 
> ```
> const firestore = getFirestore();
> useFirestoreEmulator(..., {mockUserToken});
> ```
> 
> instead of relying on `@firebase/rules-unit-testing` to create a firestore instance configured with emulator ?

Yes.

> What is the plan to update `@firebase/rules-unit-testing` the library itself and the documentation?

See http://go/firebase-rules-unit-testing-next (Googlers only). Timeline section has the remaining action items tracked.

File: packages/app-types/index.d.ts (hunk 2 above)
Comment (Feiyang1) [reply packages/app-types/index.d.ts:124]: Yes, but why do they need to be in `@firebase/app-types`?

Comment (yuchenshi) [reply packages/app-types/index.d.ts]: I see what you mean now. I've modified `@firebase/firestore-types` to peer dep on `@firebase/util` and reuse the types.

File: .changeset/lemon-ligers-protect.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+'@firebase/app-types': minor
+'@firebase/app': minor
```
Comment (Feiyang1) [inline .changeset/lemon-ligers-protect.md:3]:
```suggestion
```

File: .changeset/sweet-monkeys-warn.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"@firebase/database": patch
```
Comment (Feiyang1) [inline .changeset/sweet-monkeys-warn.md:2]: seems unrelated to this PR.

Comment (yuchenshi) [reply .changeset/sweet-monkeys-warn.md]: Good catch! Deleted

File: packages/firestore-types/index.d.ts [hunk 5]
```diff
@@ -61,7 +61,13 @@ export class FirebaseFirestore {
...
   settings(settings: Settings): void;
 
-  useEmulator(host: string, port: number): void;
+  useEmulator(
+    host: string,
+    port: number,
+    options?: {
+      mockUserToken?: EmulatorMockTokenOptions;
+    }
+  ): void;
```
Comment (schmidt-sebastian) [inline packages/firestore-types/index.d.ts:70]: This change also needs to be made in firebase/index.d.ts

File: packages/firestore/src/api/credentials.ts [hunk 6]
```diff
@@ -135,6 +135,38 @@ export class EmptyCredentialsProvider implements CredentialsProvider {
   }
 }
 
+/** A CredentialsProvider that always returns a constant token. Used for emulator token mocking. */
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/credentials.ts:138]: Nit: We pretend that we have a strict 80 character line limit.

File: packages/firestore/src/lite/database.ts [hunk 7]
```diff
@@ -247,6 +254,22 @@ export function useFirestoreEmulator(
     host: `${host}:${port}`,
     ssl: false
   });
+
+  if (options.mockUserToken) {
+    // Let createMockUserToken validate first (catches common mistakes like "uid" and missing sub.)
```
Comment (schmidt-sebastian) [inline packages/firestore/src/lite/database.ts:259]: Same comment regarding line length.

**Review by schmidt-sebastian — APPROVED** [review]

File: packages/firebase/index.d.ts [hunk 8]
```diff
@@ -15,6 +15,8 @@
  * limitations under the License.
  */
 
+import { EmulatorMockTokenOptions } from '@firebase/util';
```
Comment (Feiyang1) [inline packages/firebase/index.d.ts:18]: It is supposed to be a self-contained typing file, so should not reference other packages. Can you copy the definition here, so people can access the type using `firebase.firestore.EmulatorMockTokenOptions` ?

**Review by Feiyang1 — CHANGES_REQUESTED** [review]

File: packages/firebase/index.d.ts (hunk 8 above)
Comment (yuchenshi) [reply packages/firebase/index.d.ts:18]: Done!

File: packages/firebase/index.d.ts [hunk 9]
```diff
@@ -9976,6 +9984,81 @@ declare namespace firebase.firestore {
...
+    | 'anonymous'
+    | 'google.com'
+    | 'facebook.com'
+    | 'github.com'
+    | 'twitter.com'
+    | 'microsoft.com'
+    | 'apple.com';
+
+  interface FirebaseIdToken {
+    // Always set to https://securetoken.google.com/PROJECT_ID
```
Comment (hsubox76) [inline packages/firebase/index.d.ts:10002]: Sorry, one last comment I hope, these comments have to be jsdoc style eg
```
  /**
     * Always set to https://securetoken.google.com/PROJECT_ID
     */
```
or they won't get picked up by the doc generator. If this is supposed to be in the documentation.

Comment (yuchenshi) [reply packages/firebase/index.d.ts:10002]: Done.

**Review by hsubox76 — APPROVED** [review]

File: packages/database/src/api/Database.ts [hunk 10]
```diff
@@ -37,6 +36,14 @@ import {
 
 import { Reference } from './Reference';
 
+// TODO: revert to import {FirebaseDatabase as ExpDatabase} from '@firebase/database' once modular SDK goes GA
```
Comment (Feiyang1) [inline packages/database/src/api/Database.ts:39]: There are some unrelated changes. Can you merge or rebase with the master ?

**Review by Feiyang1 — CHANGES_REQUESTED** [review]

File: packages/database/src/api/Database.ts (hunk 10 above)
Comment (yuchenshi) [reply packages/database/src/api/Database.ts]: Rebased. These changes should be all gone now.

**Review by Feiyang1 — APPROVED** [review]
